### ISSUE TITLE
JSON Blueprints

### ISSUE THREAD

**Issue by KumoKairo** [sschmid/Entitas-CSharp#267]
I've seen a bit of talking about making a serialization system for blueprints based on Json instead of Binary Formatter.
I personally see two main problems with that:
1. Binary formatter is not very robust - if anything changes in the underlying class it breaks and doesn't deserialize anything (text formatter will at least deserialize what's already present in its fields)
2. Version control doesn't really like binary formats - one can't diff-merge a conflicting binary file, there's a need to drop one of the conflicting versions.

As there's a new modular structure of the Entitas comes with the type-safety feature, I propose the following:
- Make a Unity-only plugin of those blueprints, saving it as a plain ScriptableObject using native Unity asset serialization. JsonUtility (a built-in Json tool in Unity) makes it possible to export any ScriptableObject to a plain JSON file to send it to the server (to conduct a backend validation/whatever, most likely there's a JSON tool on a server already)
- Make a pure-C# version of the plugin, maybe leaving a current BinaryFormatter solution

I personally pretty sure that the more simple and "down-to-earth" each solution is. the better. Making a system that works everywhere makes it harder to use it in each separate case.

I've tinkered a bit with the BinaryFormatter version inside of Unity and made it save to a JSON instead, but right now it uses a nested "string" fields of the ScriptableObject to store its content. So it can be made simpler, but leaving the robustness in cases where the component IDs change after regeneration.

I need some kind of a feedback on this, especially from someone who uses Entitas with MonoGame or XNA or something

**Comment by SonicZentropy** [comment]
This is Unity-based rather than MonoGame/etc, but a while back I added an alternative JSON serializer to Entitas using  [FullSerializer](https://github.com/jacobdufault/fullserializer) and it "just worked" pretty well.  You could do the same with a different, non-Unity-centric JSON lib like JSON.Net or something and use that in MonoGame.  Highly recommend going this route of adding a new serializer rather than messing with the existing BinaryFormatter

**Comment by T2RKUS** [comment]
I couldn't seem to find a json implementation of blueprints. I know a member of the entitas community made one before but can't remember who made or where that is and believe that one could serialize references to gameobjects where as mine can not. don't have a clue where to start with that but I made one which works and thought I'd post it here in case anyone cba to write one.
[Entitas.Unity.Blueprints.Json.zip](https://github.com/sschmid/Entitas-CSharp/files/894084/Entitas.Unity.Blueprints.Json.zip)

**Comment by sschmid** [comment]
I just tested using the Unity JsonUtility instead of the BinaryFormatter again and got reminded why I went for Binary in the first place. The BinaryFormatter "just works" out of the box. Unity's JsonUtility will skip fields of type `object` but the serializable Blueprint class uses Entitas.Blueprints.SerializableMember which uses `object` for the value field. I think other Json libs can handle that but I don't want to depend on a 3rd party lib.
